ElectronClient/gui/NoteEditor/NoteBody/CodeMirror/utils/useEditorSearch.ts

```diff
@@ -0,0 +1,151 @@
+import { useCallback, useEffect, useState } from 'react';
+// Helper functions that use the cursor
+export default function useCursorUtils(CodeMirror: any) {
+
+	const [markers, setMarkers] = useState([]);
+	const [overlay, setOverlay] = useState(null);
+	const [scrollbarMarks, setScrollbarMarks] = useState(null);
+	const [overlayTimeout, setOverlayTimeout] = useState(null);
+	const [previousKeywordValue, setPreviousKeywordValue] = useState(null);
+
+	const clearMarkers = useCallback(() => {
+		for (let i = 0; i < markers.length; i++) {
+			markers[i].clear();
+		}
+
+		setMarkers([]);
+	}, [markers]);
+
+	const clearOverlay = useCallback((cm: any) => {
+		if (overlay) cm.removeOverlay(overlay);
+		if (scrollbarMarks) scrollbarMarks.clear();
+		if (overlayTimeout) clearTimeout(overlayTimeout);
+
+		setOverlay(null);
+		setScrollbarMarks(null);
+		setOverlayTimeout(null);
+	}, [overlay, scrollbarMarks, overlayTimeout]);
+
+	// Modified from codemirror/addons/search/search.js
+	function searchOverlay(query: RegExp) {
+		return { token: function(stream: any) {
+			query.lastIndex = stream.pos;
+			const match = query.exec(stream.string);
+			if (match && match.index == stream.pos) {
+				stream.pos += match[0].length || 1;
+				return 'search-marker';
+			} else if (match) {
+				stream.pos = match.index;
+			} else {
+				stream.skipToEnd();
+			}
+			return null;
+		} };
+	}
+
+	// Highlights the currently active found work
+	// It's possible to get tricky with this fucntions and just use findNext/findPrev
+	// but this is fast enough and works more naturally with the current search logic
+	function highlightSearch(cm: any, searchTerm: RegExp, index: number) {
+		const marks: any = [];
+
+		const cursor = cm.getSearchCursor(searchTerm);
+
+		let match = null;
+		for (let j = 0; j < index + 1; j++) {
+			if (!cursor.findNext()) {
+				// If we run out of matches then just highlight the final match
+				break;
+			}
+			match = cursor.pos;
+		}
+
+		if (match) {
+			marks.push(cm.markText(match.from, match.to, { className: 'cm-search-marker-selected' }));
+			cm.scrollIntoView(match);
+		}
+
+		return marks;
+	}
+
+	// https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_Expressions#Escaping
+	function escapeRegExp(keyword: string) {
+		return keyword.replace(/[.*+\-?^${}()|[\]\\]/g, '\\$&'); // $& means the whole matched string
+	}
+
+	function getSearchTerm(keyword: any) {
+		const value = escapeRegExp(keyword.value);
+		let searchTerm = new RegExp(value, 'g');
+
+		if (keyword.accuracy === 'partially') {
+			searchTerm = new RegExp(value, 'gi');
+		}
+
+		return searchTerm;
+	}
+
+	useEffect(() => {
+		return () => {
+			if (overlayTimeout) clearTimeout(overlayTimeout);
+		};
+	}, [overlayTimeout]);
```

Comment:
The logic here doesn't seem right because this effect will run every time overlayTimeout changes. In other words, as soon as you set the timeout it will be cleared here on the next render. I guess if it works now it's by accident.

Again maybe a reducer is the solution, but otherwise I've mostly been solving this using a ref: 

```js
const overlayTimeoutRef = useRef(null);
overlayTimeoutRef.current = overlayTimeout;

useEffect(() => {
	return () => {
		if (overlayTimeoutRef.current) clearTimeout(overlayTimeoutRef.current);
	};
}, []); // Only runs on unmount
```

Ugly but it works.

Reply:
Thanks for this, that's exactly what I ended up doing.